fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java

```diff
@@ -56,15 +56,14 @@ public Map<String, String> getMdcContext() {
     }
 
     static void restoreMdc(@Nonnull Map<String, String> mdcContext) {
-        MDC.clear();
-        for (Map.Entry<String, String> entry : mdcContext.entrySet()) {
-            MDC.put(entry.getKey(), entry.getValue());
-        }
+        MDC.setContextMap(mdcContext);
     }
 
     static void clearMdc(@Nonnull Map<String, String> mdcContext) {
+        Map<String, String> map = MDC.getMDCAdapter().getCopyOfContextMap();
         for (String key : mdcContext.keySet()) {
-            MDC.remove(key);
+            map.remove(key);
```

Comment:
Would it be possible/efficient to use something like: 
```
MDC.keySet().removeAll(mdcContext.keySet());
```

Reply:
no

Reply:
It is a weird implementation...

Reply:
https://stackoverflow.com/questions/17675804/remove-multiple-keys-from-map-in-efficient-way

Reply:
This is the limitation of what SLF4J 1.7 can do: https://github.com/qos-ch/slf4j/blob/e9ee55cca93c2bf26f14482a9bdf961c750d2a56/slf4j-api/src/main/java/org/slf4j/spi/MDCAdapter.java

This code is from f5e1b6d23fa18f1c9fa9bfb24ce01ce6cdc758d5 (2017-04-12), when I believe it was still log4j 1. There was some discussion then about advocating for a better API, but nothing came of it. Nothing much has changed in SLF4J 2.